src/Compilers/CSharp/Test/Emit2/Semantics/RefReadonlyParameterTests.cs

```diff
@@ -1251,4 +1251,1192 @@ static class C
+              IL_0001:  stloc.0
+              IL_0002:  ldloca.s   V_0
+              IL_0004:  call       "void C.M(ref readonly int)"
+              IL_0009:  ret
+            }
+            """);
+    }
+
+    [Fact]
+    public void RefReadonlyParameter_RValue_RefOrInArgument()
+    {
+        var source = """
+            class C
+            {
+                static void M(ref readonly int p) => System.Console.Write(p);
+                static void Main()
+                {
+                    M(ref 6);
+                    M(in 7);
+                }
+            }
+            """;
+        CreateCompilation(source).VerifyDiagnostics(
+            // (6,15): error CS1510: A ref or out value must be an assignable variable
+            //         M(ref 6);
+            Diagnostic(ErrorCode.ERR_RefLvalueExpected, "6").WithLocation(6, 15),
+            // (7,14): error CS8156: An expression cannot be used in this context because it may not be passed or returned by reference
+            //         M(in 7);
+            Diagnostic(ErrorCode.ERR_RefReturnLvalueExpected, "7").WithLocation(7, 14));
+    }
+
+    [Fact]
+    public void RefReadonlyParameter_OutArgument()
+    {
+        var source = """
+            class C
+            {
+                static void M(ref readonly int p) => System.Console.WriteLine(p);
+                static readonly int x = 5;
+                static void Main()
+                {
+                    M(out x);
+                    int y;
+                    M(out y);
+                    M(out 6);
+                }
+            }
+            """;
+        CreateCompilation(source).VerifyDiagnostics(
+            // (7,15): error CS0199: A static readonly field cannot be used as a ref or out value (except in a static constructor)
+            //         M(out x);
+            Diagnostic(ErrorCode.ERR_RefReadonlyStatic, "x").WithLocation(7, 15),
+            // (9,15): error CS1615: Argument 1 may not be passed with the 'out' keyword
+            //         M(out y);
+            Diagnostic(ErrorCode.ERR_BadArgExtraRef, "y").WithArguments("1", "out").WithLocation(9, 15),
+            // (10,15): error CS1510: A ref or out value must be an assignable variable
+            //         M(out 6);
+            Diagnostic(ErrorCode.ERR_RefLvalueExpected, "6").WithLocation(10, 15));
+    }
+
+    [Fact]
+    public void PassingParameters_In_RefReadonly_PlainArgument()
+    {
+        var source = """
+            class C
+            {
+                static void M1(in int p)
+                {
+                    M2(p);
+                }
+                static void M2(ref readonly int p) => System.Console.Write(p);
+                static void Main() => M1(5);
+            }
+            """;
+        var verifier = CompileAndVerify(source, expectedOutput: "5");
+        verifier.VerifyDiagnostics(
+            // (5,12): warning CS9503: Argument 1 should be passed with 'ref' or 'in' keyword
```

Comment:
>// (5,12): warning CS9503: Argument 1 should be passed with 'ref' or 'in' keyword [](http://example.com/codeflow?start=12&length=81)

It looks like usage of `ref` keyword is going to lead to an error: "Cannot use variable 'p' as a ref or out value because it is a readonly variable". Perhaps we shouldn't suggest `ref` for readonly variables.

Reply:
It's not, both are equally allowed per the current proposal.